Add error boundary

dcalhoun opened the pull request:
### Fix
Fixes #879. Add error boundary to provide the user a helpful message and action,
rather than an empty white page.

| Small Screens | Large Screens |
| --- | --- |
| <img width="1552" alt="error-boundary-small" src="https://user-images.githubusercontent.com/438664/110167307-31a22b80-7dbb-11eb-829f-1d550bf8a438.png"> | <img width="1552" alt="error-boundary-large" src="https://user-images.githubusercontent.com/438664/110167295-2e0ea480-7dbb-11eb-8536-67517b26944f.png"> |

| Dark Theme | Analytics Opt-In |
| --- | --- |
| <img width="1552" alt="error-boundary-large-dark" src="https://user-images.githubusercontent.com/438664/110167329-36ff7600-7dbb-11eb-8a99-23cc4c320a5b.png"> | <img width="1552" alt="error-boundary-analytics" src="https://user-images.githubusercontent.com/438664/110167314-349d1c00-7dbb-11eb-8d15-76129b90e2d1.png"> |

### Test
Testing requires introducing an intentional error to the code. After doing so,
the new error boundary should be displayed.

It is important to note that the app has two initialization paths: authenticated
and unauthenticated. To fully test, one must introduce an error in each of these
paths.

### Release
Improve error handling and messaging within the app.

dcalhoun committed:
Add error boundary

Improve user experience when a runtime error occurs by displaying a
helpful message and button to reload the application.

dcalhoun committed:
Add Electron reload event

Reloading the Electron applications requires calling
`mainWindow.reload`. This new event provides access to that method from
within the `ErrorBoundary` component.

dcalhoun committed:
Update button casing

dmsnell commented:
This is great! I'm noting that this should only catch errors in the React code or in the main UI thread, which I think are extremely rare. I was concerned about it taking over the whole page, but in those rare cases I think the page freezes or empties out.

> Rest assured that our systems have captured the details of the error so that we may work to fix the issue

I don't believe this is true. Is that something that you were planning on building into this PR as well?

dcalhoun commented:
> This is great! I'm noting that this should only catch errors in the React code or in the main UI thread, which I think are extremely rare. I was concerned about it taking over the whole page, but in those rare cases I think the page freezes or empties out.

@dmsnell just to make sure I understand correctly — you are not asking for a change here, correct? I interpreted the spirit of #879 to be to improve the "white screen of death" that occurs from errors, which is why the fullscreen message was used at the top level of the application..

> > Rest assured that our systems have captured the details of the error so that we may work to fix the issue
> 
> I don't believe this is true. Is that something that you were planning on building into this PR as well?

The copy (and the design) of the error message is definitely something I wanted to discuss with the team. So, it's not final. That said, the statement regarding "systems have captured the details" was a reference to Sentry. Based on the [existing usage of Sentry](https://github.com/Automattic/simplenote-electron/blob/da1ba4fdab26b5e6831597cb5ea241b628c88659/lib/state/analytics/middleware.ts#L63-L71) and this PRs usage of `Sentry.ErrorBoundary`, I presumed we would receive at least some information regarding any error that causes this boundary to trigger. Do you believe that to be inaccurate?

Related to the Sentry topic, I now notice that I likely need to [conditionally leverage Sentry](https://github.com/Automattic/simplenote-electron/blob/da1ba4fdab26b5e6831597cb5ea241b628c88659/lib/state/analytics/middleware.ts#L60). Is that accurate?

dmsnell commented:
>  you are not asking for a change here, correct?

that's correct. I was relating my thoughts as I processed it in case anyone else has the same ones.

> That said, the statement regarding "systems have captured the details" was a reference to Sentry.

very good. I had forgotten about that. I'm not sure if any of us are actively watching that 🙃

dmsnell commented:
> I now notice that I likely need to conditionally leverage Sentry. Is that accurate?

this is true. we should not be tracking anything about a customer without their express permission, which comes by opting in to tracking

dcalhoun commented:
Hey @SylvesterWilmott. 👋🏻 Whenever you have time, I'd appreciate your review of and input on the copy and design of this new screen. Thank you!

SylvesterWilmott commented:
Thanks @dcalhoun. Looking good!

I think we can reuse the max-width that is found in the sign up/log in screens for the layout as well as centering the content in the viewport.

I've tried to shorten the copy a little too, let me know what you think:

`Oops!`
`We’re sorry - something went wrong.
Our team has been notified so that we can work to fix the issue. Please refresh or try again later.`
`[Refresh application]`
`If the problem persists, please email us at support@simplenote.com.`

![error](https://user-images.githubusercontent.com/36532468/110125662-7af27b00-7dbb-11eb-8119-91cd1732017d.png)

dcalhoun committed:
Update copy and design

Adjust error boundary to match new design guidance.

dcalhoun committed:
Conditionally collect error data based upon user preference

If the user has not opted into analytics, we must avoid sending error
reports to Sentry.

dcalhoun committed:
Add theme support

Modify error boundary styles based upon the active theme for the user.

dcalhoun commented:
@SylvesterWilmott thank you for the guidance. I have adapted the page to match your design.

I noticed that the type and button styles for the client differ slightly from the server rendered authentication pages. So, I attempted to match your design as best I could using the existing patterns in the client app. Will you please let me know what you think of the images below? Thanks!

| Small Screens | Large Screens |
| --- | --- |
| <img width="1552" alt="error-boundary-small" src="https://user-images.githubusercontent.com/438664/110167307-31a22b80-7dbb-11eb-829f-1d550bf8a438.png"> | <img width="1552" alt="error-boundary-large" src="https://user-images.githubusercontent.com/438664/110167295-2e0ea480-7dbb-11eb-8536-67517b26944f.png"> |

| Dark Theme | Analytics Opt-In |
| --- | --- |
| <img width="1552" alt="error-boundary-large-dark" src="https://user-images.githubusercontent.com/438664/110167329-36ff7600-7dbb-11eb-8a99-23cc4c320a5b.png"> | <img width="1552" alt="error-boundary-analytics" src="https://user-images.githubusercontent.com/438664/110167314-349d1c00-7dbb-11eb-8d15-76129b90e2d1.png"> |

SylvesterWilmott commented:
Thanks @dcalhoun looks great.

Sorry for the delay here.

sandymcfadden reviewed (approved):
I wasn't familiar with Error Boundaries before this, except for a bit of reading. 
This all makes sense though and tests well.
